File: chat-app/src/main/java/gg/warcraft/chat/app/channel/GlobalChannel.java
package gg.warcraft.chat.app.channel;

import com.google.common.base.MoreObjects;
import gg.warcraft.chat.api.channel.Channel;
import gg.warcraft.monolith.api.core.event.Event;
import gg.warcraft.monolith.api.core.event.EventHandler;
import gg.warcraft.monolith.api.player.PlayerConnectEvent;
import gg.warcraft.monolith.api.player.PlayerDisconnectEvent;
import gg.warcraft.monolith.api.player.PlayerPermissionsChangedEvent;
import gg.warcraft.monolith.api.util.ColorCode;

import java.util.*;
import java.util.function.Predicate;

public class GlobalChannel implements Channel, EventHandler {
    private final String name;
    private final List<String> aliases;
    private final String shortcut;
    private final ColorCode color;
    private final String formattingString;
    private final Predicate<UUID> joinCondition;

    final Set<UUID> recipients;

    public GlobalChannel(String name, List<String> aliases, String shortcut, ColorCode color, String formattingString,
                         Predicate<UUID> joinCondition) {
        this.name = name;
        this.aliases = aliases;
        this.shortcut = shortcut;
        this.color = color;
        this.formattingString = formattingString;
        this.joinCondition = joinCondition;
        this.recipients = new HashSet<>();
    }

    @Override
    public String getName() {
        return name;
    }

    @Override
    public List<String> getAliases() {
        return aliases;
    }

    @Override
    public String getShortcut() {
        return shortcut;
    }

    @Override
    public ColorCode getColor() {
        return color;
    }

    @Override
    public String getFormattingString() {
        return formattingString;
    }

    public List<UUID> getRecipients() {
        return new ArrayList<>(recipients);
    }

    public Predicate<UUID> getJoinCondition() {
        return joinCondition;
    }

    // TODO move event handling out of channel object?
    @Override
    public void handle(Event event) {
        if (event instanceof PlayerConnectEvent) {
            onPlayerConnect((PlayerConnectEvent) event);
        } else if (event instanceof PlayerDisconnectEvent) {
            onPlayerDisconnect((PlayerDisconnectEvent) event);
        } else if (event instanceof PlayerPermissionsChangedEvent) {
            onPlayerPermissionsChanged((PlayerPermissionsChangedEvent) event);
        }
    }

    public void onPlayerConnect(PlayerConnectEvent event) {
        if (joinCondition.test(event.playerId())) {
            recipients.add(event.playerId());
        }
    }

    public void onPlayerPermissionsChanged(PlayerPermissionsChangedEvent event) {
        if (recipients.remove(event.playerId())) {
            if (joinCondition.test(event.playerId())) {
                recipients.add(event.playerId());
            }
        }
    }

    public void onPlayerDisconnect(PlayerDisconnectEvent event) {
        recipients.remove(event.playerId());
    }

    @Override
    public String toString() {
        return MoreObjects.toStringHelper(getClass())
                .add("name", name)
                .add("aliases", aliases)
                .add("shortcut", shortcut)
                .add("color", color)
                .add("formattingString", formattingString)
                .toString();
    }
}

File: chat/src/main/scala/gg/warcraft/chat/channel/GlobalChannel.scala
package gg.warcraft.chat.channel

import java.util.UUID

import gg.warcraft.chat.profile.ChatProfileService
import gg.warcraft.chat.{Message, MessageAdapter}
import gg.warcraft.monolith.api.core.command.{Command, CommandSender}
import gg.warcraft.monolith.api.core.event.{Event, EventHandler}
import gg.warcraft.monolith.api.entity.player.Player
import gg.warcraft.monolith.api.entity.player.service.PlayerQueryService
import gg.warcraft.monolith.api.player.{
  PlayerConnectEvent, PlayerDisconnectEvent, PlayerPermissionsChangedEvent
}
import gg.warcraft.monolith.api.util.ColorCode

import scala.collection.mutable

case class GlobalChannel(
    name: String,
    aliases: Set[String],
    shortcut: Option[String],
    color: ColorCode,
    formatString: String,
    permission: Option[String]
)(
    private implicit val playerService: PlayerQueryService,
    override protected implicit val profileService: ChatProfileService,
    override protected implicit val messageAdapter: MessageAdapter
) extends Channel
    with EventHandler {
  private final val missingPermissions =
    "You do not have the required permissions to join [%s]."
  private final val successfullyJoined =
    "You have joined [%s]."
  private final val homeChannelPlayersOnly =
    "Only players can set their home channel."

  private val recipients = mutable.ListBuffer[UUID]()

  override def handle(sender: CommandSender, cmd: Command): Boolean = sender match {
    case CommandSender(_, Some(playerId)) =>
      val joined = if (!recipients.contains(playerId)) {
        permission match {
          case Some(permission) =>
            val player: Player = playerService.getPlayer(playerId)
            if (player.hasPermission(permission)) {
              recipients += playerId
              val message = Message.server(successfullyJoined.format(name))
              // messageCommandService.sendMessageToPlayer(joinedMessage, sender.playerId().get());
              true
            } else {
              val message = Message.server(missingPermissions.format(name))
              // messageCommandService.sendMessageToPlayer(missingPermissionsMessage, sender.playerId().get());
              false
            }
          case _ =>
            recipients += playerId
            val message = Message.server(successfullyJoined.format(name))
            // messageCommandService.sendMessageToPlayer(joinedMessage, sender.playerId().get());
            true
        }
      } else true

      if (joined) {
        if (cmd.args.isEmpty) makeHome(playerId)
        else broadcast(sender, cmd.args.mkString(" "), recipients)
      }

      true

    case _ =>
      if (cmd.args.isEmpty) println(homeChannelPlayersOnly)
      else broadcast(sender, cmd.args.mkString(" "), recipients)
      true
  }

  override def handle(event: Event): Unit = event match {
    case PlayerConnectEvent(playerId, _) =>
      permission match {
        case Some(permission) =>
          val player: Player = playerService.getPlayer(playerId)
          if (player.hasPermission(permission)) recipients += playerId
        case _ => recipients += playerId
      }

    case PlayerPermissionsChangedEvent(playerId, _) =>
      permission match {
        case Some(permission) =>
          val player: Player = playerService.getPlayer(playerId)
          if (recipients.contains(playerId)) {
            if (!player.hasPermission(permission)) recipients -= playerId
          } else {
            if (player.hasPermission(permission)) recipients += playerId
          }
        case _ => ()
      }

    case PlayerDisconnectEvent(playerId, _) =>
      permission match {
        case Some(permission) =>
          val player: Player = playerService.getPlayer(playerId)
          if (player.hasPermission(permission)) recipients -= playerId
        case _ => recipients -= playerId
      }

    case _ => ()
  }
}
